module/common.go

```diff
@@ -30,3 +36,98 @@ func (n *NoopReadDoneAware) Done() <-chan struct{} {
 	defer close(done)
 	return done
 }
+
+// Startable provides an interface to start a component. Once started, the component
+// can be stopped by cancelling the given context.
+type Startable interface {
+	Start(irrecoverable.SignalerContext) error
+}
+
+type Component interface {
+	Startable
+	ReadyDoneAware
+}
+
+type ComponentFactory func() (Component, error)
+
+// OnError reacts to an irrecoverable error
+// It could:
+// - restart the component (in production) after cleanup
+// - panic (in canary / benchmark)
+// - log in various Error channels and / or send telemetry ...
+type OnError = func(err error, triggerRestart func())
+
+func RunComponent(ctx context.Context, componentFactory ComponentFactory, handler OnError) error {
+	// reference to per-run signals for the component
+	var component Component
+	var cancel context.CancelFunc
+	var done <-chan struct{}
+	var irrecoverables chan error
+
+	start := func() error {
+		var err error // startup error, should be handled out of band
+
+		component, err = componentFactory()
+		if err != nil {
+			return err // failure to generate the component, should be handles out-of-band because a restart won't help
+		}
+
+		// context used to run the component
+		var runCtx context.Context
+		runCtx, cancel = context.WithCancel(ctx)
+
+		// signaler used for irrecoverables
+		var signalingCtx irrecoverable.SignalerContext
+		irrecoverables = make(chan error)
+		signalingCtx = irrecoverable.WithSignaler(runCtx, irrecoverable.NewSignaler(irrecoverables))
+
+		if err = component.Start(signalingCtx); err != nil {
+			// failed to start component: this should not trigger a restart
+			return err
+		}
+
+		// wait for Ready
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-component.Ready():
+		}
+
+		done = component.Done()
+		return nil
+	}
+
+	for {
+		if err := start(); err != nil {
+			return err // failure to start
+		}
+
+		select {
+		case err := <-irrecoverables:
+			// shutdown the component
+			cancel()
+
+			// wait until it's done
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-done:
+			}
+
+			// send error to the handler programmed with a restart continuation
+			restartChan := make(chan struct{})
+			go handler(err, func() {
+				close(restartChan)
+			})
+
+			// wait for handler to trigger restart or abort
```

Comment:
With the `func` we're passing as the second argument to the handler, how do we expect the handler to perform this restart?